README.md
cosi2
=====


cosi/History

July 1  2011   Big changes:

         Added "sentinel" method of detecting regions that have coalesced,
         replacing original dg_done_coalescent() code.

			Changed how mutations are placed on the genealogy: instead of tracing each
			individual mutation down the genealogy, for each segment of each seglist
			we compute the set of leaves that get any mutations carried by that segment.
			These leafsets are constructed incrementally as new seglists are constructed.
			This replaces the sim_mutate() step.

			Changed the simulator to keep only the currently active nodes (those currently
			in a Pop) and place mutations as we go along, instead of first constructing
			the entire ARG and then placing mutations on it.  This reduces memory usage
			and improves data locality.

         Changed Seglist representation to be a flat vector of
			[ beg_0 end_0 beg_1 end_1 ... beg_k end_k ].

			Small changes:

         Changed Seglist representation to be a flat vector of
			[ beg_0 end_0 beg_1 end_1 ... beg_k end_k ].  (This lets us
			quickly check if a location falls on a segment or in a gap,
			by doing one binary search and checking the parity of the
			resulting index).

			Changed the remove operation on nodelists to run in constant time:
			removed element is now replaced with the last element of the array,
			instead of shifting all elements above the removed element.
			This is ok since the order of nodes within a population is unimportant.

June 08 2011	Remove segment.h/c (replaced by seglist). Add one "break" when getting
     		head node. V06-07-11a.

June 07 2011	Turn segments into arrays as well; no output change. V06-07-11a.
     		(Adds seglist.h/c)

June 01 2011	Turn list of recomb sites into arrays. Still no output change. 
     		Tag as V06-01-01a.

May 31 2011	Get rid of definition of rsitenum (= int). Tagged as V05-31-11a.

May 26 2011	Tidying things up in a number of places: Turn poplist into array, and 
       		eliminate poplist arguments from 2 dg_ functions (getting pop by name). 
		No longer remobe migrations to/from non-existent pops Do not add 
		migration if rate is zero (Historical.c deletes previous migration, and 
		new one has no effect.) Have migration hold pop pointers, rather than 
		names. Change dg_coalesce_by_name to dg_coalesce_by_pop. No change to 
		output.  Tag as V05-26-11a.

May 17 2011	Add protection to sweep.c, so that will not end sweep with nsel=2 (only happens for 
       		large s).

May 6 2011	One more header problem that I left behind. Removed multiplying random seed by -1, to 
      		make compatible with transition to mtwist in cosi2_rand (rile.c). Tag as V05-06-11a. 
		(Correction: some things weren't tagged right. Tag as V05-10-11a.)

May 4 2011	Cleaned up various header-related issues. No functional changes. Tag as V05-04-11a.

May 3 2011	Remove obsolete copyright notice. Add some $\ID lines. 
      		Update to new directory for cosi2_rand header files. Added nested-header protetion to 
		some .h files.  Tag as V05-03-11a.

Apr 12 2011	Import from cosi for renewed development. Tag as V04-12-11a.

=================================================================================================
Nov 08 2010	Make error output more specific when invalid parameter is encountered in reading 
       		file. Not tagged, but included in 

Jul 19 2010	Swap indices in sweep_execute(), when deciding which gene conversion chunk has the 
       		parental allele. The existing code was backwards (i.e. wrong), as Ilya discovered.

		Tag as V071910a

Jun 9 2007	Add line to geneconversion.c so that overall rate is calculated both when length 
		and when rate set. Otherwise, specifying length param first => no gene conversion.

Mar 29 2007	Remove test in sweep.c that ends sweep when down to single selected chrom, since 
		recombination onto selected allele could still occur (should be minor effect).

Nov 16 2006	Remove debugging output from sweep.c. Fix error in sweep:sw_recomb_pick_popindex, 
		where I was incorrectly staggering random counter by 0.5, leading to possibility 
		of picking population with zero members.

Nov 15 2006	Modify sweep to output selected allele. Tag as V080906a.
		New param format: 
		pop_event sweep "selective sweep" 5[pop] 10000[time] .02[s] .5[pos] .4[final freq]

Oct 19 2006	-g becomes -O3 in Makefile.

		Tag as V080906a.

Jan 27 2006	Uncovered modest conceptual bug in sweep.c: was summing probability of event over 
		time intervals until passed random threshold. This is wrong, since the sum will
		deterministically exceed 1.0, while the probability of nothing happens is always 
		> 0.  Instead, calc only prob in current time slice and toss random number every slice.

Jan 24 2006	Working on sweep with other populations present. Revise simulator.c to always 
		check whether done after historical event, and change some function names
		in demography.c to make more consistent interface.

Jan 17 2006	Code modifications for speed. Now keep track of how many active nodes there
		are for each location (i.e. for each segment), to reduce checking.
		Also tried maintaining list of nodes associated with each segment (see forays/demography.stage2), 
		but ran slower under most circumstances.
		V011806a.

Sep 22 2005	Add copyright.  V09-22-05a.  Replace as Release 1.0.

Sep 16 2005	Release 1.0.

June 21 2005	Bug fix in bottleneck.c -- require at least 2 nodes to attempt bottleneck execution.  V06-21-6a.

April 26 2005 	Removed increment to time after bottleneck (historical.c), since it was screwing up historical events
unless spaced by >= 1 gen.  V04-26-05a.

Feb 4 2005	Tract length change somehow failed to get into tagged version (but was tranferred to public_html, etc)
Retag as V02-04-05a.

Nov 5 2004	Gene conversion tract length set to 500 bp.  V11-05-04a.  Moving this to public_html area.

Oct 19 2004	Switched to having start end of partial sweep be number of chromosomes. V10-19-04a.


cosi/file.cc
/* $Id: file.c,v 1.4 2011/05/06 15:35:43 sfs Exp $ */
#include <cstdio>
#include <cstring>
#include <cstdlib>
#include <cctype>
#include <cerrno>
#include <sstream>
#include <fstream>
#include <boost/make_shared.hpp>
#include <boost/algorithm/string.hpp>
#include <boost/filesystem.hpp>
#include <boost/filesystem/fstream.hpp>
#include <boost/exception/all.hpp>
#include <boost/algorithm/string/trim.hpp>
#include <cosi/file.h>
#include <cosi/recomb.h>
#include <cosi/demography.h>
#include <cosi/simulator.h>
#include <cosi/output.h>
#include <cosi/geneconversion.h>
#include <cosi/genmap.h>
#include <cosi/historical.h>
#include <cosi/sweep.h>
#include <cosi/utils.h>

namespace cosi {

struct cosi_param_file_error: virtual cosi_io_error {};
typedef boost::error_info<struct errinfo_err_detail_,std::string> errinfo_err_detail;
typedef boost::error_info<struct errinfo_file_name2_, filename_t> errinfo_file_name2;
typedef boost::error_info<struct errinfo_param_name_,std::string> errinfo_param_name;
typedef boost::error_info<struct errinfo_param_details_,std::string> errinfo_param_details;

static const size_t BUF_MAX = 1024;

ParamFileReader::ParamFileReader( DemographyP demography_ ):
	demography( demography_ ), genMapShift( 0 ) {
	init();
}

void ParamFileReader::init() {
  seeded = False;
  popsize = 0;
  sampsize = 0;
  length = 0;
  mu = prob_per_bp_per_gen_t( 0.0 );
  recombrate = 0;
	geneConv2RecombRateRatio = factor_t(0.0);
	geneConversionMeanTractLength = 500 /* bp */;
	geneConversionMinTractLength = 4 /* bp */;
	geneConversionModel = GeneConversion::GCM_LEFT_AND_RIGHT_FROM_ORIGIN;
  rseed = 0;

	histEvents = boost::make_shared<HistEvents>( demography );
	infSites = False;
	printSeed = True;
	ignoreRecombsInPop = NULL_POPID;
}


void ParamFileReader::file_read(boost::filesystem::path filename, FILE *segfp) 
{
	init();
	this->paramFileName = filename;

	try {
		FILE *infileptr = cosi_fopen (filename.c_str(), "r");
		if ( !infileptr ) {
			BOOST_THROW_EXCEPTION( cosi_param_file_error()
														 << boost::errinfo_errno(errno)
														 << error_msg( "Error opening param file" ) );
		}
		if (segfp != NULL) {
		  std::string s = filename.string();
		  fprintf(segfp, "paramfile: %s\n", s.c_str());
		}
		file_get_data (infileptr, segfp);
		fclose(infileptr);
	} catch( boost::exception& e ) {
		e << boost::errinfo_file_name( filename.string() );
		throw;
	}
}

/*********************************************************/

int 
ParamFileReader::file_proc_buff(char *var, char* buffer, FILE* segfp) 
{

	popid     popname;
	int  intarg;
		
	/*	printf("%s\n", var); */
	if (strcmp(var,"length") == 0) {
		length = atoi(buffer);
	}
	else if (strcmp(var,"recomb_file") == 0) {
		if (buffer[strlen(buffer)-1] == '\n')
			 buffer[strlen(buffer)-1] = '\0';
		chkCond( length > 0, "cosi paramfile error: must specify length before recombmap" );

		namespace fs = boost::filesystem;

		fs::path recombFilePath( buffer );
		if ( !this->recombfileFN.empty() ) recombFilePath = this->recombfileFN;
		fs::path paramFileDir( fs::canonical( this->paramFileName ).parent_path() );
		recombFilePath = fs::absolute( recombFilePath /*, paramFileDir */ );
		
		genMap = boost::make_shared<GenMap>( recombFilePath, length, this->genMapShift );
		assert( genMap.get() );
	}
	else if (strcmp(var, "mutation_rate") == 0) {
		mu = prob_per_bp_per_gen_t( atof(buffer) );
	}
	else if (strcmp(var, "infinite_sites") == 0) {
	  if (buffer[strlen(buffer)-1] == '\n')
			 buffer[strlen(buffer)-1] = '\0';
	  if (strcmp(buffer, "yes") == 0 || strcmp(buffer, "Yes") == 0 || strcmp(buffer, "YES") == 0)
			 infSites = True;
	}
	else if (strcmp(var, "gene_conversion_rate") == 0) {
		//geneConversionRate = atof(buffer);
		file_exit( "deprecated_parameter", "please specify gene_conversion_relative_rate giving ratio of gene conversion to recomb rate" );
	}
	else if (strcmp(var, "gene_conversion_relative_rate") == 0) {
		geneConv2RecombRateRatio = factor_t( atof(buffer) );
	}
	else if (strcmp(var, "gene_conversion_mean_tract_length") == 0) {
		geneConversionMeanTractLength = atoi(buffer);
	}
	else if (strcmp(var, "gene_conversion_min_tract_length") == 0) {
		geneConversionMinTractLength = atoi(buffer);
	}
	else if (strcmp(var, "gene_conversion_model") == 0) {
		geneConversionModel = GeneConversion::parseGCModel( boost::trim_copy( string( buffer ) ) );
	}
	else if (strcmp(var, "pop_size") == 0) {
		popname = popid( atoi(strtok (buffer, " ")) );
		intarg = atoi(strtok (NULL, " " ));
		if (FILE_DEBUG)
			 printf("popsize: %d\n", intarg);
		/* 
		 * Throw a fatal error if pop [popname] does not exist.
		 */		
		if (! demography->dg_set_pop_size_by_name (ZERO_GEN, popname, intarg))
			 file_exit("file_proc_buff", 
								 "parameter file - pop specified does not exist.");
	}
	else if (strcmp(var, "sample_size") == 0) {		
		popname = popid( atoi(strtok (buffer, " ")) );
		intarg = atoi(strtok (NULL, " " ));
		if (FILE_DEBUG)
			 printf("sampsize: %d\n", intarg);
		demography->dg_populate_by_name (popname, intarg);
		if (segfp != NULL) {fprintf(segfp, "A %d %d\n", ToInt( popname ), intarg);}
	}
	else if ( !strcmp( var, "pop_ignore_recombs" ) ) {
		this->ignoreRecombsInPop = popid( atoi(strtok (buffer, " ")) );
	}
	else if (strcmp(var, "pop_define") == 0) {
		popname = popid( atoi(strtok (buffer, " ")) );
		string pop_label( strtok (NULL, " ") );
		if ( pop_label.empty() ) pop_label = "some_pop";
		if ( pop_label.at( pop_label.size()-1 ) == '\n' )
			 pop_label.resize( pop_label.size()-1 );
		demography->dg_create_pop (popname, pop_label, /* genid=*/ ZERO_GEN);
	}
	else if (strcmp(var, "pop_event") == 0) {
		histEvents->addEvent( histEvents->parseEvent( buffer ) );
	}
	else if (strcmp(var, "random_seed") == 0) {
	  std::istringstream is( buffer );
	  rseed = -1;
	  is >> rseed;
	  if (rseed > 0) {
	    //set_rng_seed(rseed);
//	    if ( printSeed ) std::cerr << "coalescent seed set in param file:" << rseed << "\n";
	    seeded = True;
	  } else
			 cerr << "cosi error -- could not read seed: " << buffer << endl;
	} else if (strcmp(var,"sweep_traj_file") == 0) {
		if (buffer[strlen(buffer)-1] == '\n')
			 buffer[strlen(buffer)-1] = '\0';
		//sweep_set_traj_file(buffer);
	} else if (!strcmp(var, "pop_traj_file" )) {
		std::istringstream is( buffer );
		is.exceptions( std::ios_base::failbit | std::ios_base::badbit );
		filename_t sizeTrajFN;
		popid popname( NULL_POPID );
			
		try { is >> popname >> sizeTrajFN; }
		catch( const std::ios_base::failure& e ) {
			BOOST_THROW_EXCEPTION( cosi_param_file_error()
														 << error_msg( "invalid pop size traj line" )
														 << boost::errinfo_errno( errno ) );
		}
		try {
			math::Function< genid, popsize_float_t, math::Piecewise< math::Const<> > > sizeTraj;
			boost::filesystem::ifstream sizeTrajFile;
			try { sizeTrajFile.open( sizeTrajFN ); }
			catch( const std::ios_base::failure& e ) {
				BOOST_THROW_EXCEPTION( cosi_param_file_error()
															 << error_msg( "could not open pop size traj file" )
															 << boost::errinfo_errno( errno ) );
			}
			loadFrom( sizeTrajFile, sizeTraj );
			this->pop2sizeTraj.insert( std::make_pair( popname, sizeTraj ) );
		} catch( boost::exception& e ) {
			e << errinfo_file_name2( sizeTrajFN );
			throw;
		}
	} else {
		BOOST_THROW_EXCEPTION( cosi_param_file_error() << error_msg( "unknown param file line" ) );
	}

	return 1;
}

int 
ParamFileReader::file_get_data (FILE *fileptr, FILE *segfp) 
{
	char c,
		 buffer[BUF_MAX],
		 var[50];
	unsigned lineNum = 0;

	c = getc(fileptr);
	while (c != EOF) {
		switch (c) {
		case '#':
			c = getc(fileptr);
			while (c != '\n' && c != EOF)
				 c = getc(fileptr);
			lineNum++;
			break;
		case '\n':
			c = getc(fileptr);
			lineNum++;
			break;
		case ' ':
			c = getc(fileptr);
			break;
		default:
			ungetc(c, fileptr);
			fscanf(fileptr, "%s", var);
			file_killwhitespace(fileptr);
			fgets(buffer, BUF_MAX, fileptr);
			try {
				try { file_proc_buff(var, buffer, segfp); }
				catch( const std::ios_base::failure& e ) {
					BOOST_THROW_EXCEPTION( cosi_param_file_error()
																 << boost::errinfo_errno( errno ) );
				}
			} catch( boost::exception& e ) {
				e << boost::errinfo_at_line( lineNum )
					<< errinfo_param_name( var )
					<< errinfo_param_details( boost::trim_copy( std::string( buffer ) ) );
				throw;
			}
			c = getc(fileptr);
			lineNum++;
			break;
		}
	}

	// if (!seeded)
	// 	 std::cerr << "coalescent seed: " << ( rseed = seed_rng()) << endl;

	if (segfp != NULL) {fprintf(segfp, "params: length %d mu %.9f\n", length, (double)ToDouble( mu ) );}
	if (segfp != NULL) {fprintf(segfp, "L %d\n", length);}

	return 1;
}

int 
ParamFileReader::file_killwhitespace(FILE * fileptr) 
{
	char c;
	int i = 0;
	c = getc(fileptr);
	while (isspace((int) c)) {
		c = getc(fileptr);
		i++;
	}
	ungetc(c, fileptr);
	return i;
}

/*
 * FILE_EXIT
 */
void 
ParamFileReader::file_exit(const char* funct_name, const char* error_string) 
{
	fprintf(stderr, "file.c | %s: %s\n",
					funct_name, error_string);
	exit(EXIT_FAILURE);
}

void 
ParamFileReader::file_error_nonfatal(const char* funct_name, const char* error_string) 
{
	fprintf(stderr, "file.c | %s: %s\n",
					funct_name, error_string);
}

}  // namespace cosi
